tests/unit/artifact/test_repository.py

```diff
@@ -313,3 +317,190 @@ def test_parse_rpm_string_valid(pkg_string, expected):
+        _ = provider.artifacts
+
+
+@patch('tmt.steps.prepare.artifact.providers.tmt.utils.retry_session')
+def test_repository_provider_fetch_contents(mock_retry_session, root_logger):
+    """Test fetch_contents method discovers RPMs from repository"""
+
+    # Mock the Repository.from_url call
+    mock_session = MagicMock()
+    mock_response = MagicMock()
+    mock_response.text = VALID_REPO_CONTENT
+    mock_response.raise_for_status.return_value = None
+    mock_session.get.return_value = mock_response
+    mock_retry_session.return_value.__enter__.return_value = mock_session
+
+    # Mock guest and package manager
+    mock_guest = MagicMock()
+    mock_package_manager = MagicMock()
+    mock_guest.package_manager = mock_package_manager
+
+    # Mock list_packages to return some RPMs
+    mock_package_manager.list_packages.return_value = [
+        "docker-ce-1:20.10.7-3.el8.x86_64",
+        "docker-ce-cli-1:20.10.7-3.el8.x86_64",
+        "containerd.io-1.4.6-3.1.el8.x86_64",
+    ]
+
+    provider = RepositoryFileProvider(
+        "repository-url:https://download.docker.com/linux/centos/docker-ce.repo", root_logger
+    )
+
+    # Call fetch_contents
+    result = provider.fetch_contents(mock_guest, Path("/tmp/artifacts"))
+
+    # Verify result is empty list (discovery-only provider)
+    assert result == []
+
+    # Verify package manager methods were called
+    mock_package_manager.install_repository.assert_called_once()
+    mock_package_manager.list_packages.assert_called_once()
+
+    # Verify artifacts property now works
+    artifacts = provider.artifacts
+    assert len(artifacts) == 3
+    assert artifacts[0]._raw_artifact["name"] == "docker-ce"
+    assert artifacts[0]._raw_artifact["version"] == "20.10.7"
+    assert artifacts[0]._raw_artifact["epoch"] == "1"
+    assert artifacts[0]._raw_artifact["release"] == "3.el8"
+    assert artifacts[0]._raw_artifact["arch"] == "x86_64"
+
+
+@patch('tmt.steps.prepare.artifact.providers.tmt.utils.retry_session')
+def test_repository_provider_malformed_packages(mock_retry_session, root_logger):
+    """Test that malformed package strings are skipped with warnings"""
+
+    # Mock the Repository.from_url call
+    mock_session = MagicMock()
+    mock_response = MagicMock()
+    mock_response.text = VALID_REPO_CONTENT
+    mock_response.raise_for_status.return_value = None
+    mock_session.get.return_value = mock_response
+    mock_retry_session.return_value.__enter__.return_value = mock_session
+
+    # Mock guest and package manager
+    mock_guest = MagicMock()
+    mock_package_manager = MagicMock()
+    mock_guest.package_manager = mock_package_manager
+
+    # Mock list_packages with mix of valid and malformed packages
+    mock_package_manager.list_packages.return_value = [
+        "docker-ce-1:20.10.7-3.el8.x86_64",  # Valid
+        "invalid-package-string",  # Invalid - no arch
+        "bash-5.1.8-6.el9.x86_64",  # Valid
+        "another-malformed",  # Invalid
```

Comment:
Catch the log for the invalid ones. But really, if ever this thing fails, we *really* need to know the context of why this thing failed.